### PR TITLE
Improve performance of compute_colliding_points

### PR DESCRIPTION
**Pull request by jorgensd**
This branch:
```bash
Build tree (mus): 8901 16510 10080
Coll points (mus): 569879 592972  582874
Coll tree (mus): 7398 10531  9247
Closest (mus): 4999853 5526093  5194065
```
current flatten geometry:
```bash
Build tree (mus): 8862 14524 10244
Coll points (mus): 828933 1038671  918395
Coll tree (mus): 11113 15795  12441
Closest (mus): 5272091 6084143  5674116
```
main
```bash
Build tree (mus): 9078 12151 9917
Coll points (mus): 836218 924306  859247
Coll tree (mus): 11612 16093  12789
Closest (mus): 5577778 6991559  6276822
```
Script: (Note one line has to change between pr and branches)
```c++

#include <basix/finite-element.h>
#include <chrono>
#include <dolfinx.h>
#include <dolfinx/fem/FiniteElement.h>
#include <dolfinx/geometry/utils.h>

using namespace dolfinx;
int main(int argc, char* argv[])
{
  dolfinx::init_logging(argc, argv);
  MPI_Init(&argc, &argv);

  {
    // Create mesh and function space
    auto part = mesh::create_cell_partitioner(mesh::GhostMode::shared_facet);
    auto mesh = std::make_shared<mesh::Mesh>(
        mesh::create_box(MPI_COMM_WORLD, {{{0.0, 0.0, 0.0}, {1.0, 1.0, 1.0}}},
                         {15, 15, 15}, mesh::CellType::tetrahedron, part));

    const std::size_t num_runs = 10;
    std::vector<int> build_time(num_runs);
    std::vector<int> coll_time(num_runs);
    for (std::size_t i = 0; i < num_runs; ++i)
    {
      auto begint = std::chrono::high_resolution_clock::now();
      geometry::BoundingBoxTree tree(*mesh, mesh->topology().dim());
      auto endt = std::chrono::high_resolution_clock::now();
      auto durationt
          = std::chrono::duration_cast<std::chrono::microseconds>(endt - begint)
                .count();
      build_time[i] = durationt;
      basix::FiniteElement b_el = basix::create_element(
          basix::element::family::P, basix::cell::type::tetrahedron, 3,
          basix::element::lagrange_variant::equispaced);
      fem::FiniteElement el(b_el, 1);
      std::size_t num_cells_local
          = mesh->topology().index_map(mesh->topology().dim())->size_local();
      std::vector<std::int32_t> cells(num_cells_local);
      std::iota(cells.begin(), cells.end(), 0);
      std::vector<double> coords = fem::interpolation_coords(el, *mesh, cells);

      auto begin = std::chrono::high_resolution_clock::now();
      graph::AdjacencyList<std::int32_t> cols
          = dolfinx::geometry::compute_collisions(
              tree, std::span<const double>(coords));
      auto end = std::chrono::high_resolution_clock::now();

      auto duration
          = std::chrono::duration_cast<std::chrono::microseconds>(end - begin)
                .count();
      coll_time[i] = duration;
    }

    {
      auto min_time = std::min_element(build_time.begin(), build_time.end());
      auto max_time = std::max_element(build_time.begin(), build_time.end());
      auto avg_time = std::accumulate(build_time.begin(), build_time.end(), 0);
      std::cout << "Build tree (mus): " << *min_time << " " << *max_time << " "
                << avg_time / num_runs << std::endl;
    }

    {
      auto min_time = std::min_element(coll_time.begin(), coll_time.end());
      auto max_time = std::max_element(coll_time.begin(), coll_time.end());
      auto avg_time = std::accumulate(coll_time.begin(), coll_time.end(), 0);
      std::cout << "Coll points (mus): " << *min_time << " " << *max_time << " "
                << " " << avg_time / num_runs << std::endl;
    }

    {
      auto mesh_hex = std::make_shared<mesh::Mesh>(
          mesh::create_box(MPI_COMM_WORLD, {{{0.3, 0.4, 0.2}, {1.3, 1.2, 1.8}}},
                           {15, 15, 15}, mesh::CellType::hexahedron, part));
      geometry::BoundingBoxTree tree(*mesh, mesh->topology().dim());
      geometry::BoundingBoxTree hex_tree(*mesh_hex, mesh_hex->topology().dim());
      std::vector<double> hex_coll(num_runs);
      for (std::size_t i = 0; i < num_runs; ++i)
      {
        auto begint = std::chrono::high_resolution_clock::now();
        // main
        std::vector<std::array<int, 2>> colls
            = geometry::compute_collisions(tree, hex_tree);
        // branches
        //  std::vector<int> colls = geometry::compute_collisions(tree,
        //  hex_tree);
        auto endt = std::chrono::high_resolution_clock::now();
        auto durationt = std::chrono::duration_cast<std::chrono::microseconds>(
                             endt - begint)
                             .count();
        hex_coll[i] = durationt;
      }
      auto min_time = std::min_element(hex_coll.begin(), hex_coll.end());
      auto max_time = std::max_element(hex_coll.begin(), hex_coll.end());
      auto avg_time = std::accumulate(hex_coll.begin(), hex_coll.end(), 0);
      std::cout << "Coll tree (mus): " << *min_time << " " << *max_time << " "
                << " " << avg_time / num_runs << std::endl;
    }

    {
      std::vector<int> closest_time(num_runs);
      geometry::BoundingBoxTree tree(*mesh, mesh->topology().dim());
      basix::FiniteElement b_el = basix::create_element(
          basix::element::family::P, basix::cell::type::tetrahedron, 3,
          basix::element::lagrange_variant::equispaced);
      fem::FiniteElement el(b_el, 1);
      std::size_t num_cells_local
          = mesh->topology().index_map(mesh->topology().dim())->size_local();
      std::vector<std::int32_t> cells(num_cells_local);
      std::iota(cells.begin(), cells.end(), 0);
      std::vector<double> coords = fem::interpolation_coords(el, *mesh, cells);
      auto midpoint_tree = geometry::create_midpoint_tree(
          *mesh, mesh->topology().dim(), cells);
      for (std::size_t i = 0; i < num_runs; ++i)
      {
        auto begin = std::chrono::high_resolution_clock::now();
        std::vector<int> entities = dolfinx::geometry::compute_closest_entity(
            tree, midpoint_tree, *mesh, std::span<const double>(coords));
        auto end = std::chrono::high_resolution_clock::now();

        auto duration
            = std::chrono::duration_cast<std::chrono::microseconds>(end - begin)
                  .count();
        closest_time[i] = duration;
      }
      auto min_time
          = std::min_element(closest_time.begin(), closest_time.end());
      auto max_time
          = std::max_element(closest_time.begin(), closest_time.end());
      auto avg_time
          = std::accumulate(closest_time.begin(), closest_time.end(), 0);
      std::cout << "Closest (mus): " << *min_time << " " << *max_time << " "
                << " " << avg_time / num_runs << std::endl;
    }
  }
  MPI_Finalize();

  return 0;
}
```

### CODE REVIEW

File: cpp/dolfinx/geometry/BoundingBoxTree.h [hunk 1]
```diff
@@ -71,7 +71,13 @@ class BoundingBoxTree
   /// Return bounding box coordinates for a given node in the tree
   /// @param[in] node The bounding box node index
   /// @return A copy of bounding box (lower_corner, upper_corner). Shape (2,3).
-  std::array<double, 6> get_bbox(std::size_t node) const;
+  std::array<double, 6> copy_bbox(std::size_t node) const;
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/BoundingBoxTree.h:74]: I don't think this is need and it's redundant. The caller can copy the span into and array if they want a copy.

File: python/dolfinx/wrappers/geometry.cpp [hunk 2]
```diff
@@ -241,7 +241,7 @@ void geometry(py::module& m)
           [](const dolfinx::geometry::BoundingBoxTree& self,
              const std::size_t i)
           {
```
Comment (garth-wells) [inline python/dolfinx/wrappers/geometry.cpp:243]: This function is more complicated than it needs to be. Just create a `py::array_t` from `bbox` pointer and size (6). By default pybind11 will copy the data.

**Review by garth-wells — CHANGES_REQUESTED** [review]
Which specific change/line leads to the performance change?

**Comment by jorgensd** [comment]
Superseeded by #2359

**Pull request closed without merging**